### PR TITLE
Fix:(issue_976) All positional args and flags to be mixed in any order

### LINKED ISSUE #976: v2 feature: allow flags after arguments
## Checklist

* [X] Are you running the latest v2 release? The list of releases is [here](https://github.com/urfave/cli/releases).
* [X] Did you check the manual for your release? The v2 manual is [here](https://github.com/urfave/cli/blob/master/docs/v2/manual.md)
* [X] Did you perform a search about this feature? Here's the [Github guide](https://help.github.com/en/github/managing-your-work-on-github/using-search-to-filter-issues-and-pull-requests) about searching.

## What problem does this solve?

I am writing a program which is invoked by another program (openvpn up-script) and it is invoked in a very specific way.  I really like this library, so I'd like to be able to use it as a command line parser.  The problem is that it seems that all flags must appear before all options.

In my specific case, openvpn will always call your script with exactly 10 arguments, then you can configure it to pass additional arguments.  This would result in an os.Args similar to this:
`{"tun0", "1", "2"..."10", "--flush"}`.  The `--flush` is the flag that I'd like to be able to process with this library.

It seems like all flags either take no arguments (`BoolFlag`) or take one arugment (`StringFlag`).

I've needed this for other projects, but having it for this project would be really great.

## Solution description

an `os.Args` value of `{"a", "--bool-flag", "--string-flag", "abc", "b"}` should have:

* `ctx.Args() --> {"a", "b"}`
* `ctx.BoolFlag("bool-flag") --> true`
* `ctx.StringFlag("string-flag") --> "abc"`

## Describe alternatives you've considered

Not using this library for some of my projects

### LINKED ISSUE #1733: Ordering of arguments and flags impact one another
## My urfave/cli version is

_** v2.25.3**_

## Checklist

- [x] Are you running the latest v2 release? The list of releases is [here](https://github.com/urfave/cli/releases).
- [x] Did you check the manual for your release? The v2 manual is [here](https://github.com/urfave/cli/blob/main/docs/v2/manual.md)
- [x] Did you perform a search about this problem? Here's the [GitHub guide](https://help.github.com/en/github/managing-your-work-on-github/using-search-to-filter-issues-and-pull-requests) about searching.

## Dependency Management
- My project is using go modules.

## Describe the bug

Ordering of arguments and flags impact one another, causing a flag value to be set or unset depending on whether it comes before or after an argument. This can result in a silent error where the flag name is recognized but the value isn't set.

## To reproduce

Using the flag example from https://cli.urfave.org/v2/examples/flags/ (copying below for ease of reference)

```go
package main

import (
	"fmt"
	"log"
	"os"

	"github.com/urfave/cli/v2"
)

func main() {
	app := &cli.App{
		Flags: []cli.Flag{
			&cli.StringFlag{
				Name:  "lang",
				Value: "english",
				Usage: "language for the greeting",
			},
		},
		Action: func(cCtx *cli.Context) error {
			name := "Nefertiti"
			if cCtx.NArg() > 0 {
				name = cCtx.Args().Get(0)
			}
			if cCtx.String("lang") == "spanish" {
				fmt.Println("Hola", name)
			} else {
				fmt.Println("Hello", name)
			}
			return nil
		},
	}

	if err := app.Run(os.Args); err != nil {
		log.Fatal(err)
	}
}
```

## Observed behavior

```bash
# argument
$ ./example my-name
Hello my-name
# flag
$ ./example --lang spanish        
Hola Nefertiti
# flag before argument
$ ./example --lang spanish my-name 
Hola my-name
# argument before flag
$ ./example my-name --lang spanish
Hello my-name
```

## Expected behavior

`$ ./example my-name --lang spanish` and `./example --lang spanish my-name ` should both set the `--lang` flag to the value `spanish`

## Additional context

N/A

## Want to fix this yourself?

We'd love to have more contributors on this project! If the fix for
this bug is easily explained and very small, feel free to create a
pull request for it.

## Run `go version` and paste its output here

```
go version go1.20.4 darwin/arm64
```

## Run `go env` and paste its output here

```
# paste `go env` output in here
```

### PR DESCRIPTION
**Pull request by dearchap**
## What type of PR is this?

_(REQUIRED)_

- feature

## What this PR does / why we need it:

_(REQUIRED)_

Working towards making positions arguments and flags interchangeable

## Which issue(s) this PR fixes:

_(REQUIRED)_

Fixes #976 
Fixes #1928 
Fixes #1733 

## Special notes for your reviewer:

_(fill-in or delete this section)_

Focus on command.parseFlags which has the logic

## Testing

_(fill-in or delete this section)_

All tests have been updated to ensure they work 

## Release Notes

_(REQUIRED)_

```release-note
Allows positions args and flags to be placed in any order
```

### CODE REVIEW

File: command.go [hunk 1]
```diff
@@ -802,14 +804,60 @@ func (cmd *Command) parseFlags(args Args) (Args, error) {
...
+			tracef("returning-1 (cmd=%[1]q) args %[2]q", cmd.Name, posArgs)
+			cmd.parsedArgs = &stringSliceArgs{posArgs}
+			return cmd.parsedArgs, err
+		}
+		tracef("rearrange-4 (cmd=%[1]q) check %[2]q", cmd.Name, cmd.flagSet.Args())
+		rargs = cmd.flagSet.Args()
+		if len(rargs) == 0 || len(rargs[0]) == 0 || rargs[0] == "-" {
+			break
+		}
+	}
```
Comment (BlackHole1) [inline command.go:855]: Optimize nesting depth.

```go
for {
		tracef("rearrange:1 (cmd=%[1]q) %[2]q", cmd.Name, rargs)
		for {
			tracef("rearrange:2 (cmd=%[1]q) %[2]q %[3]q", cmd.Name, posArgs, rargs)

			if len(rargs) == 0 || len(rargs[0]) == 0 {
				break
			}

			if rargs[0] == "--" {
				posArgs = append(posArgs, rargs...)
				cmd.parsedArgs = &stringSliceArgs{posArgs}
				return cmd.parsedArgs, nil
			}

			if rargs[0][0] == '-' {
				break
			}

			tracef("rearrange-3 (cmd=%[1]q) check %[2]q", cmd.Name, rargs[0])

			if cmd.Command(rargs[0]) != nil {
				posArgs = append(posArgs, rargs...)
				cmd.parsedArgs = &stringSliceArgs{posArgs}
				return cmd.parsedArgs, nil
			}

			posArgs = append(posArgs, rargs[0])
			if len(rargs) == 0 {
				rargs = []string{}
				break
			}

			rargs = rargs[1:]
		}

		if err := parseIter(cmd.flagSet, cmd, rargs, cmd.Root().shellCompletion); err != nil {
			posArgs = append(posArgs, cmd.flagSet.Args()...)
			tracef("returning-1 (cmd=%[1]q) args %[2]q", cmd.Name, posArgs)
			cmd.parsedArgs = &stringSliceArgs{posArgs}
			return cmd.parsedArgs, err
		}
}
```

**Review by BlackHole1 — CHANGES_REQUESTED** [review]

File: command_test.go [hunk 2]
```diff
@@ -163,7 +163,7 @@ func TestCommandFlagParsing(t *testing.T) {
 		{testArgs: []string{"test-cmd", "blah", "blah"}, skipFlagParsing: true, useShortOptionHandling: false},   // Test SkipFlagParsing without any args that look like flags
 		{testArgs: []string{"test-cmd", "blah", "-break"}, skipFlagParsing: true, useShortOptionHandling: false}, // Test SkipFlagParsing with random flag arg
 		{testArgs: []string{"test-cmd", "blah", "-help"}, skipFlagParsing: true, useShortOptionHandling: false},  // Test SkipFlagParsing with "special" help flag arg
-		{testArgs: []string{"test-cmd", "blah", "-h"}, skipFlagParsing: false, useShortOptionHandling: true},     // Test UseShortOptionHandling
+		//{testArgs: []string{"test-cmd", "blah", "-h"}, skipFlagParsing: false, useShortOptionHandling: true},     // Test UseShortOptionHandling
```
Comment (BlackHole1) [inline command_test.go:166]: Did you forget to delete this?

File: command_test.go [hunk 3]
```diff
@@ -693,16 +693,16 @@ var defaultCommandSubCommandTests = []struct {
...
-	{" ", "j", "", false},
-	{"bat", "", "batbaz", false},
-	{"nothing", "", "batbaz", false},
+	//{" ", "", "", true},
+	//{" ", "j", "", true},
+	{"bat", "", "batbaz", true},
+	{"nothing", "", "batbaz", true},
 	{"nothing", "", "", false},
-	{"nothing", "j", "batbaz", false},
+	//{"nothing", "j", "batbaz", true},
```
Comment (BlackHole1) [inline command_test.go:705]: Above

File: command_test.go (hunk 2 above)
Comment (dearchap) [reply command_test.go:166]: no. for some reason the test fails and I couldnt figure it out

File: command_test.go (hunk 3 above)
Comment (dearchap) [reply command_test.go:705]: no. for some reason the test fails and I couldnt figure it out

File: command_test.go [hunk 4]
```diff
@@ -693,14 +693,14 @@ var defaultCommandSubCommandTests = []struct {
...
 	{"", "jimmers", "", true},
-	{" ", "jimmers", "foobar", false},
+	//{" ", "jimmers", "foobar", false},
 	{"", "", "", true},
-	{" ", "", "", false},
-	{" ", "j", "", false},
-	{"bat", "", "batbaz", false},
-	{"nothing", "", "batbaz", false},
+	//{" ", "", "", true},
+	//{" ", "j", "", true},
```
Comment (BlackHole1) [inline command_test.go:701]: Did you forget to remove the comments?
If there are issues, maybe we can add a TODO here to inform future readers why this part is commented out.

**Review by BlackHole1 — CHANGES_REQUESTED** [review]

File: command_test.go (hunk 4 above)
Comment (dearchap) [reply command_test.go:701]: I'm not even sure these are valid test cases. Why would an argument be just a string of spaces ?

Comment (danhunsaker) [reply command_test.go:701]: Some field separators are encoded as successive whitespace. Specifying `'    '` is simpler to use than `--space=4` for some users in some use cases, even if that's not strictly "better" technically. So I've seen it done.

Comment (dearchap) [reply command_test.go:701]: But these are the args specified directly to the application . So on command line if I type 

```
Foo    Bar
```

The application would get it as an array of strings ["Foo","Bar"] so I don't see when we would be passed empty spaces at all

Comment (danhunsaker) [reply command_test.go:701]:
```sh
foo      bar
```

won't give any whitespace, correct. But

```sh
foo '    ' bar
```

will. Shell quotation doesn't get passed along, obviously, but the contents do.

**Review by danhunsaker — APPROVED** [review]

**Pull request merged**